{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Install libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(biomaRt)\n",
    "library(org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_ids <- c(\"NR_075077\",\"NM_001005337\",\"NM_012102\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I am showing ways to convert refseq ids to gene names using `org.Hs.en.db` and `biomaRt`. Based on the recovered refseq ID, it conclusively shows that `org.Hs.en.db` is the better option for refSeq and `biomaRt` for Ensembl. \n",
    "\n",
    "https://www.bioconductor.org/help/course-materials/2015/UseBioconductorFeb2015/A01.5_Annotation.html\n",
    "\n",
    "# Using org.Hs.eg.db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ACCNUM'</li><li>'ALIAS'</li><li>'ENSEMBL'</li><li>'ENSEMBLPROT'</li><li>'ENSEMBLTRANS'</li><li>'ENTREZID'</li><li>'ENZYME'</li><li>'EVIDENCE'</li><li>'EVIDENCEALL'</li><li>'GENENAME'</li><li>'GENETYPE'</li><li>'GO'</li><li>'GOALL'</li><li>'IPI'</li><li>'MAP'</li><li>'OMIM'</li><li>'ONTOLOGY'</li><li>'ONTOLOGYALL'</li><li>'PATH'</li><li>'PFAM'</li><li>'PMID'</li><li>'PROSITE'</li><li>'REFSEQ'</li><li>'SYMBOL'</li><li>'UCSCKG'</li><li>'UNIPROT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ACCNUM'\n",
       "\\item 'ALIAS'\n",
       "\\item 'ENSEMBL'\n",
       "\\item 'ENSEMBLPROT'\n",
       "\\item 'ENSEMBLTRANS'\n",
       "\\item 'ENTREZID'\n",
       "\\item 'ENZYME'\n",
       "\\item 'EVIDENCE'\n",
       "\\item 'EVIDENCEALL'\n",
       "\\item 'GENENAME'\n",
       "\\item 'GENETYPE'\n",
       "\\item 'GO'\n",
       "\\item 'GOALL'\n",
       "\\item 'IPI'\n",
       "\\item 'MAP'\n",
       "\\item 'OMIM'\n",
       "\\item 'ONTOLOGY'\n",
       "\\item 'ONTOLOGYALL'\n",
       "\\item 'PATH'\n",
       "\\item 'PFAM'\n",
       "\\item 'PMID'\n",
       "\\item 'PROSITE'\n",
       "\\item 'REFSEQ'\n",
       "\\item 'SYMBOL'\n",
       "\\item 'UCSCKG'\n",
       "\\item 'UNIPROT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ACCNUM'\n",
       "2. 'ALIAS'\n",
       "3. 'ENSEMBL'\n",
       "4. 'ENSEMBLPROT'\n",
       "5. 'ENSEMBLTRANS'\n",
       "6. 'ENTREZID'\n",
       "7. 'ENZYME'\n",
       "8. 'EVIDENCE'\n",
       "9. 'EVIDENCEALL'\n",
       "10. 'GENENAME'\n",
       "11. 'GENETYPE'\n",
       "12. 'GO'\n",
       "13. 'GOALL'\n",
       "14. 'IPI'\n",
       "15. 'MAP'\n",
       "16. 'OMIM'\n",
       "17. 'ONTOLOGY'\n",
       "18. 'ONTOLOGYALL'\n",
       "19. 'PATH'\n",
       "20. 'PFAM'\n",
       "21. 'PMID'\n",
       "22. 'PROSITE'\n",
       "23. 'REFSEQ'\n",
       "24. 'SYMBOL'\n",
       "25. 'UCSCKG'\n",
       "26. 'UNIPROT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"ACCNUM\"       \"ALIAS\"        \"ENSEMBL\"      \"ENSEMBLPROT\"  \"ENSEMBLTRANS\"\n",
       " [6] \"ENTREZID\"     \"ENZYME\"       \"EVIDENCE\"     \"EVIDENCEALL\"  \"GENENAME\"    \n",
       "[11] \"GENETYPE\"     \"GO\"           \"GOALL\"        \"IPI\"          \"MAP\"         \n",
       "[16] \"OMIM\"         \"ONTOLOGY\"     \"ONTOLOGYALL\"  \"PATH\"         \"PFAM\"        \n",
       "[21] \"PMID\"         \"PROSITE\"      \"REFSEQ\"       \"SYMBOL\"       \"UCSCKG\"      \n",
       "[26] \"UNIPROT\"     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "keytypes(org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ACCNUM'</li><li>'ALIAS'</li><li>'ENSEMBL'</li><li>'ENSEMBLPROT'</li><li>'ENSEMBLTRANS'</li><li>'ENTREZID'</li><li>'ENZYME'</li><li>'EVIDENCE'</li><li>'EVIDENCEALL'</li><li>'GENENAME'</li><li>'GENETYPE'</li><li>'GO'</li><li>'GOALL'</li><li>'IPI'</li><li>'MAP'</li><li>'OMIM'</li><li>'ONTOLOGY'</li><li>'ONTOLOGYALL'</li><li>'PATH'</li><li>'PFAM'</li><li>'PMID'</li><li>'PROSITE'</li><li>'REFSEQ'</li><li>'SYMBOL'</li><li>'UCSCKG'</li><li>'UNIPROT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ACCNUM'\n",
       "\\item 'ALIAS'\n",
       "\\item 'ENSEMBL'\n",
       "\\item 'ENSEMBLPROT'\n",
       "\\item 'ENSEMBLTRANS'\n",
       "\\item 'ENTREZID'\n",
       "\\item 'ENZYME'\n",
       "\\item 'EVIDENCE'\n",
       "\\item 'EVIDENCEALL'\n",
       "\\item 'GENENAME'\n",
       "\\item 'GENETYPE'\n",
       "\\item 'GO'\n",
       "\\item 'GOALL'\n",
       "\\item 'IPI'\n",
       "\\item 'MAP'\n",
       "\\item 'OMIM'\n",
       "\\item 'ONTOLOGY'\n",
       "\\item 'ONTOLOGYALL'\n",
       "\\item 'PATH'\n",
       "\\item 'PFAM'\n",
       "\\item 'PMID'\n",
       "\\item 'PROSITE'\n",
       "\\item 'REFSEQ'\n",
       "\\item 'SYMBOL'\n",
       "\\item 'UCSCKG'\n",
       "\\item 'UNIPROT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ACCNUM'\n",
       "2. 'ALIAS'\n",
       "3. 'ENSEMBL'\n",
       "4. 'ENSEMBLPROT'\n",
       "5. 'ENSEMBLTRANS'\n",
       "6. 'ENTREZID'\n",
       "7. 'ENZYME'\n",
       "8. 'EVIDENCE'\n",
       "9. 'EVIDENCEALL'\n",
       "10. 'GENENAME'\n",
       "11. 'GENETYPE'\n",
       "12. 'GO'\n",
       "13. 'GOALL'\n",
       "14. 'IPI'\n",
       "15. 'MAP'\n",
       "16. 'OMIM'\n",
       "17. 'ONTOLOGY'\n",
       "18. 'ONTOLOGYALL'\n",
       "19. 'PATH'\n",
       "20. 'PFAM'\n",
       "21. 'PMID'\n",
       "22. 'PROSITE'\n",
       "23. 'REFSEQ'\n",
       "24. 'SYMBOL'\n",
       "25. 'UCSCKG'\n",
       "26. 'UNIPROT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"ACCNUM\"       \"ALIAS\"        \"ENSEMBL\"      \"ENSEMBLPROT\"  \"ENSEMBLTRANS\"\n",
       " [6] \"ENTREZID\"     \"ENZYME\"       \"EVIDENCE\"     \"EVIDENCEALL\"  \"GENENAME\"    \n",
       "[11] \"GENETYPE\"     \"GO\"           \"GOALL\"        \"IPI\"          \"MAP\"         \n",
       "[16] \"OMIM\"         \"ONTOLOGY\"     \"ONTOLOGYALL\"  \"PATH\"         \"PFAM\"        \n",
       "[21] \"PMID\"         \"PROSITE\"      \"REFSEQ\"       \"SYMBOL\"       \"UCSCKG\"      \n",
       "[26] \"UNIPROT\"     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "columns(org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>REFSEQ</th><th scope=col>SYMBOL</th><th scope=col>GENENAME</th><th scope=col>ENSEMBL</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>NR_075077   </td><td>C1orf141</td><td>chromosome 1 open reading frame 141     </td><td>ENSG00000203963</td></tr>\n",
       "\t<tr><td>NM_001005337</td><td>PKP1    </td><td>plakophilin 1                           </td><td>ENSG00000081277</td></tr>\n",
       "\t<tr><td>NM_012102   </td><td>RERE    </td><td>arginine-glutamic acid dipeptide repeats</td><td>ENSG00000142599</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 4\n",
       "\\begin{tabular}{llll}\n",
       " REFSEQ & SYMBOL & GENENAME & ENSEMBL\\\\\n",
       " <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t NR\\_075077    & C1orf141 & chromosome 1 open reading frame 141      & ENSG00000203963\\\\\n",
       "\t NM\\_001005337 & PKP1     & plakophilin 1                            & ENSG00000081277\\\\\n",
       "\t NM\\_012102    & RERE     & arginine-glutamic acid dipeptide repeats & ENSG00000142599\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 4\n",
       "\n",
       "| REFSEQ &lt;chr&gt; | SYMBOL &lt;chr&gt; | GENENAME &lt;chr&gt; | ENSEMBL &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| NR_075077    | C1orf141 | chromosome 1 open reading frame 141      | ENSG00000203963 |\n",
       "| NM_001005337 | PKP1     | plakophilin 1                            | ENSG00000081277 |\n",
       "| NM_012102    | RERE     | arginine-glutamic acid dipeptide repeats | ENSG00000142599 |\n",
       "\n"
      ],
      "text/plain": [
       "  REFSEQ       SYMBOL   GENENAME                                \n",
       "1 NR_075077    C1orf141 chromosome 1 open reading frame 141     \n",
       "2 NM_001005337 PKP1     plakophilin 1                           \n",
       "3 NM_012102    RERE     arginine-glutamic acid dipeptide repeats\n",
       "  ENSEMBL        \n",
       "1 ENSG00000203963\n",
       "2 ENSG00000081277\n",
       "3 ENSG00000142599"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cols <- c(\"SYMBOL\", \"GENENAME\", \"ENSEMBL\")\n",
    "\n",
    "select(org.Hs.eg.db, keys=refseq_ids, \n",
    "       columns=cols, keytype=\"REFSEQ\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using biomaRt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "##using biomaRt, assigned gene ids to refseq ids\n",
    "ensembl_mart <- useDataset(\"hsapiens_gene_ensembl\", useMart(\"ensembl\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3157 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>name</th><th scope=col>description</th><th scope=col>page</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ensembl_gene_id                </td><td>Gene stable ID                              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_gene_id_version        </td><td>Gene stable ID version                      </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id          </td><td>Transcript stable ID                        </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id_version  </td><td>Transcript stable ID version                </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id             </td><td>Protein stable ID                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id_version     </td><td>Protein stable ID version                   </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_exon_id                </td><td>Exon stable ID                              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>description                    </td><td>Gene description                            </td><td>feature_page</td></tr>\n",
       "\t<tr><td>chromosome_name                </td><td>Chromosome/scaffold name                    </td><td>feature_page</td></tr>\n",
       "\t<tr><td>start_position                 </td><td>Gene start (bp)                             </td><td>feature_page</td></tr>\n",
       "\t<tr><td>end_position                   </td><td>Gene end (bp)                               </td><td>feature_page</td></tr>\n",
       "\t<tr><td>strand                         </td><td>Strand                                      </td><td>feature_page</td></tr>\n",
       "\t<tr><td>band                           </td><td>Karyotype band                              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_start               </td><td>Transcript start (bp)                       </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_end                 </td><td>Transcript end (bp)                         </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcription_start_site       </td><td>Transcription start site (TSS)              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_length              </td><td>Transcript length (including UTRs and CDS)  </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_tsl                 </td><td>Transcript support level (TSL)              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_gencode_basic       </td><td>GENCODE basic annotation                    </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_appris              </td><td>APPRIS annotation                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_is_canonical        </td><td>Ensembl Canonical                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_mane_select         </td><td>RefSeq match transcript (MANE Select)       </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_mane_plus_clinical  </td><td>RefSeq match transcript (MANE Plus Clinical)</td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_gene_name             </td><td>Gene name                                   </td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_gene_source           </td><td>Source of gene name                         </td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_transcript_name       </td><td>Transcript name                             </td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_transcript_source_name</td><td>Source of transcript name                   </td><td>feature_page</td></tr>\n",
       "\t<tr><td>percentage_gene_gc_content     </td><td>Gene % GC content                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>gene_biotype                   </td><td>Gene type                                   </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_biotype             </td><td>Transcript type                             </td><td>feature_page</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>5_utr_end                    </td><td>5' UTR end                                </td><td>sequences</td></tr>\n",
       "\t<tr><td>3_utr_start                  </td><td>3' UTR start                              </td><td>sequences</td></tr>\n",
       "\t<tr><td>3_utr_end                    </td><td>3' UTR end                                </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id        </td><td>Transcript stable ID                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id_version</td><td>Transcript stable ID version              </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id           </td><td>Protein stable ID                         </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id_version   </td><td>Protein stable ID version                 </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_biotype           </td><td>Transcript type                           </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_version           </td><td>Version (transcript)                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>peptide_version              </td><td>Version (protein)                         </td><td>sequences</td></tr>\n",
       "\t<tr><td>strand                       </td><td>Strand                                    </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_start             </td><td>Transcript start (bp)                     </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_end               </td><td>Transcript end (bp)                       </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcription_start_site     </td><td>Transcription start site (TSS)            </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_length            </td><td>Transcript length (including UTRs and CDS)</td><td>sequences</td></tr>\n",
       "\t<tr><td>cds_length                   </td><td>CDS Length                                </td><td>sequences</td></tr>\n",
       "\t<tr><td>cds_start                    </td><td>CDS start                                 </td><td>sequences</td></tr>\n",
       "\t<tr><td>cds_end                      </td><td>CDS end                                   </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_exon_id              </td><td>Exon stable ID                            </td><td>sequences</td></tr>\n",
       "\t<tr><td>exon_chrom_start             </td><td>Exon region start (bp)                    </td><td>sequences</td></tr>\n",
       "\t<tr><td>exon_chrom_end               </td><td>Exon region end (bp)                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>strand                       </td><td>Strand                                    </td><td>sequences</td></tr>\n",
       "\t<tr><td>rank                         </td><td>Exon rank in transcript                   </td><td>sequences</td></tr>\n",
       "\t<tr><td>phase                        </td><td>Start phase                               </td><td>sequences</td></tr>\n",
       "\t<tr><td>end_phase                    </td><td>End phase                                 </td><td>sequences</td></tr>\n",
       "\t<tr><td>cdna_coding_start            </td><td>cDNA coding start                         </td><td>sequences</td></tr>\n",
       "\t<tr><td>cdna_coding_end              </td><td>cDNA coding end                           </td><td>sequences</td></tr>\n",
       "\t<tr><td>genomic_coding_start         </td><td>Genomic coding start                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>genomic_coding_end           </td><td>Genomic coding end                        </td><td>sequences</td></tr>\n",
       "\t<tr><td>is_constitutive              </td><td>Constitutive exon                         </td><td>sequences</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3157 × 3\n",
       "\\begin{tabular}{lll}\n",
       " name & description & page\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ensembl\\_gene\\_id                 & Gene stable ID                               & feature\\_page\\\\\n",
       "\t ensembl\\_gene\\_id\\_version         & Gene stable ID version                       & feature\\_page\\\\\n",
       "\t ensembl\\_transcript\\_id           & Transcript stable ID                         & feature\\_page\\\\\n",
       "\t ensembl\\_transcript\\_id\\_version   & Transcript stable ID version                 & feature\\_page\\\\\n",
       "\t ensembl\\_peptide\\_id              & Protein stable ID                            & feature\\_page\\\\\n",
       "\t ensembl\\_peptide\\_id\\_version      & Protein stable ID version                    & feature\\_page\\\\\n",
       "\t ensembl\\_exon\\_id                 & Exon stable ID                               & feature\\_page\\\\\n",
       "\t description                     & Gene description                             & feature\\_page\\\\\n",
       "\t chromosome\\_name                 & Chromosome/scaffold name                     & feature\\_page\\\\\n",
       "\t start\\_position                  & Gene start (bp)                              & feature\\_page\\\\\n",
       "\t end\\_position                    & Gene end (bp)                                & feature\\_page\\\\\n",
       "\t strand                          & Strand                                       & feature\\_page\\\\\n",
       "\t band                            & Karyotype band                               & feature\\_page\\\\\n",
       "\t transcript\\_start                & Transcript start (bp)                        & feature\\_page\\\\\n",
       "\t transcript\\_end                  & Transcript end (bp)                          & feature\\_page\\\\\n",
       "\t transcription\\_start\\_site        & Transcription start site (TSS)               & feature\\_page\\\\\n",
       "\t transcript\\_length               & Transcript length (including UTRs and CDS)   & feature\\_page\\\\\n",
       "\t transcript\\_tsl                  & Transcript support level (TSL)               & feature\\_page\\\\\n",
       "\t transcript\\_gencode\\_basic        & GENCODE basic annotation                     & feature\\_page\\\\\n",
       "\t transcript\\_appris               & APPRIS annotation                            & feature\\_page\\\\\n",
       "\t transcript\\_is\\_canonical         & Ensembl Canonical                            & feature\\_page\\\\\n",
       "\t transcript\\_mane\\_select          & RefSeq match transcript (MANE Select)        & feature\\_page\\\\\n",
       "\t transcript\\_mane\\_plus\\_clinical   & RefSeq match transcript (MANE Plus Clinical) & feature\\_page\\\\\n",
       "\t external\\_gene\\_name              & Gene name                                    & feature\\_page\\\\\n",
       "\t external\\_gene\\_source            & Source of gene name                          & feature\\_page\\\\\n",
       "\t external\\_transcript\\_name        & Transcript name                              & feature\\_page\\\\\n",
       "\t external\\_transcript\\_source\\_name & Source of transcript name                    & feature\\_page\\\\\n",
       "\t percentage\\_gene\\_gc\\_content      & Gene \\% GC content                            & feature\\_page\\\\\n",
       "\t gene\\_biotype                    & Gene type                                    & feature\\_page\\\\\n",
       "\t transcript\\_biotype              & Transcript type                              & feature\\_page\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t 5\\_utr\\_end                     & 5' UTR end                                 & sequences\\\\\n",
       "\t 3\\_utr\\_start                   & 3' UTR start                               & sequences\\\\\n",
       "\t 3\\_utr\\_end                     & 3' UTR end                                 & sequences\\\\\n",
       "\t ensembl\\_transcript\\_id         & Transcript stable ID                       & sequences\\\\\n",
       "\t ensembl\\_transcript\\_id\\_version & Transcript stable ID version               & sequences\\\\\n",
       "\t ensembl\\_peptide\\_id            & Protein stable ID                          & sequences\\\\\n",
       "\t ensembl\\_peptide\\_id\\_version    & Protein stable ID version                  & sequences\\\\\n",
       "\t transcript\\_biotype            & Transcript type                            & sequences\\\\\n",
       "\t transcript\\_version            & Version (transcript)                       & sequences\\\\\n",
       "\t peptide\\_version               & Version (protein)                          & sequences\\\\\n",
       "\t strand                        & Strand                                     & sequences\\\\\n",
       "\t transcript\\_start              & Transcript start (bp)                      & sequences\\\\\n",
       "\t transcript\\_end                & Transcript end (bp)                        & sequences\\\\\n",
       "\t transcription\\_start\\_site      & Transcription start site (TSS)             & sequences\\\\\n",
       "\t transcript\\_length             & Transcript length (including UTRs and CDS) & sequences\\\\\n",
       "\t cds\\_length                    & CDS Length                                 & sequences\\\\\n",
       "\t cds\\_start                     & CDS start                                  & sequences\\\\\n",
       "\t cds\\_end                       & CDS end                                    & sequences\\\\\n",
       "\t ensembl\\_exon\\_id               & Exon stable ID                             & sequences\\\\\n",
       "\t exon\\_chrom\\_start              & Exon region start (bp)                     & sequences\\\\\n",
       "\t exon\\_chrom\\_end                & Exon region end (bp)                       & sequences\\\\\n",
       "\t strand                        & Strand                                     & sequences\\\\\n",
       "\t rank                          & Exon rank in transcript                    & sequences\\\\\n",
       "\t phase                         & Start phase                                & sequences\\\\\n",
       "\t end\\_phase                     & End phase                                  & sequences\\\\\n",
       "\t cdna\\_coding\\_start             & cDNA coding start                          & sequences\\\\\n",
       "\t cdna\\_coding\\_end               & cDNA coding end                            & sequences\\\\\n",
       "\t genomic\\_coding\\_start          & Genomic coding start                       & sequences\\\\\n",
       "\t genomic\\_coding\\_end            & Genomic coding end                         & sequences\\\\\n",
       "\t is\\_constitutive               & Constitutive exon                          & sequences\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3157 × 3\n",
       "\n",
       "| name &lt;chr&gt; | description &lt;chr&gt; | page &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| ensembl_gene_id                 | Gene stable ID                               | feature_page |\n",
       "| ensembl_gene_id_version         | Gene stable ID version                       | feature_page |\n",
       "| ensembl_transcript_id           | Transcript stable ID                         | feature_page |\n",
       "| ensembl_transcript_id_version   | Transcript stable ID version                 | feature_page |\n",
       "| ensembl_peptide_id              | Protein stable ID                            | feature_page |\n",
       "| ensembl_peptide_id_version      | Protein stable ID version                    | feature_page |\n",
       "| ensembl_exon_id                 | Exon stable ID                               | feature_page |\n",
       "| description                     | Gene description                             | feature_page |\n",
       "| chromosome_name                 | Chromosome/scaffold name                     | feature_page |\n",
       "| start_position                  | Gene start (bp)                              | feature_page |\n",
       "| end_position                    | Gene end (bp)                                | feature_page |\n",
       "| strand                          | Strand                                       | feature_page |\n",
       "| band                            | Karyotype band                               | feature_page |\n",
       "| transcript_start                | Transcript start (bp)                        | feature_page |\n",
       "| transcript_end                  | Transcript end (bp)                          | feature_page |\n",
       "| transcription_start_site        | Transcription start site (TSS)               | feature_page |\n",
       "| transcript_length               | Transcript length (including UTRs and CDS)   | feature_page |\n",
       "| transcript_tsl                  | Transcript support level (TSL)               | feature_page |\n",
       "| transcript_gencode_basic        | GENCODE basic annotation                     | feature_page |\n",
       "| transcript_appris               | APPRIS annotation                            | feature_page |\n",
       "| transcript_is_canonical         | Ensembl Canonical                            | feature_page |\n",
       "| transcript_mane_select          | RefSeq match transcript (MANE Select)        | feature_page |\n",
       "| transcript_mane_plus_clinical   | RefSeq match transcript (MANE Plus Clinical) | feature_page |\n",
       "| external_gene_name              | Gene name                                    | feature_page |\n",
       "| external_gene_source            | Source of gene name                          | feature_page |\n",
       "| external_transcript_name        | Transcript name                              | feature_page |\n",
       "| external_transcript_source_name | Source of transcript name                    | feature_page |\n",
       "| percentage_gene_gc_content      | Gene % GC content                            | feature_page |\n",
       "| gene_biotype                    | Gene type                                    | feature_page |\n",
       "| transcript_biotype              | Transcript type                              | feature_page |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| 5_utr_end                     | 5' UTR end                                 | sequences |\n",
       "| 3_utr_start                   | 3' UTR start                               | sequences |\n",
       "| 3_utr_end                     | 3' UTR end                                 | sequences |\n",
       "| ensembl_transcript_id         | Transcript stable ID                       | sequences |\n",
       "| ensembl_transcript_id_version | Transcript stable ID version               | sequences |\n",
       "| ensembl_peptide_id            | Protein stable ID                          | sequences |\n",
       "| ensembl_peptide_id_version    | Protein stable ID version                  | sequences |\n",
       "| transcript_biotype            | Transcript type                            | sequences |\n",
       "| transcript_version            | Version (transcript)                       | sequences |\n",
       "| peptide_version               | Version (protein)                          | sequences |\n",
       "| strand                        | Strand                                     | sequences |\n",
       "| transcript_start              | Transcript start (bp)                      | sequences |\n",
       "| transcript_end                | Transcript end (bp)                        | sequences |\n",
       "| transcription_start_site      | Transcription start site (TSS)             | sequences |\n",
       "| transcript_length             | Transcript length (including UTRs and CDS) | sequences |\n",
       "| cds_length                    | CDS Length                                 | sequences |\n",
       "| cds_start                     | CDS start                                  | sequences |\n",
       "| cds_end                       | CDS end                                    | sequences |\n",
       "| ensembl_exon_id               | Exon stable ID                             | sequences |\n",
       "| exon_chrom_start              | Exon region start (bp)                     | sequences |\n",
       "| exon_chrom_end                | Exon region end (bp)                       | sequences |\n",
       "| strand                        | Strand                                     | sequences |\n",
       "| rank                          | Exon rank in transcript                    | sequences |\n",
       "| phase                         | Start phase                                | sequences |\n",
       "| end_phase                     | End phase                                  | sequences |\n",
       "| cdna_coding_start             | cDNA coding start                          | sequences |\n",
       "| cdna_coding_end               | cDNA coding end                            | sequences |\n",
       "| genomic_coding_start          | Genomic coding start                       | sequences |\n",
       "| genomic_coding_end            | Genomic coding end                         | sequences |\n",
       "| is_constitutive               | Constitutive exon                          | sequences |\n",
       "\n"
      ],
      "text/plain": [
       "     name                           \n",
       "1    ensembl_gene_id                \n",
       "2    ensembl_gene_id_version        \n",
       "3    ensembl_transcript_id          \n",
       "4    ensembl_transcript_id_version  \n",
       "5    ensembl_peptide_id             \n",
       "6    ensembl_peptide_id_version     \n",
       "7    ensembl_exon_id                \n",
       "8    description                    \n",
       "9    chromosome_name                \n",
       "10   start_position                 \n",
       "11   end_position                   \n",
       "12   strand                         \n",
       "13   band                           \n",
       "14   transcript_start               \n",
       "15   transcript_end                 \n",
       "16   transcription_start_site       \n",
       "17   transcript_length              \n",
       "18   transcript_tsl                 \n",
       "19   transcript_gencode_basic       \n",
       "20   transcript_appris              \n",
       "21   transcript_is_canonical        \n",
       "22   transcript_mane_select         \n",
       "23   transcript_mane_plus_clinical  \n",
       "24   external_gene_name             \n",
       "25   external_gene_source           \n",
       "26   external_transcript_name       \n",
       "27   external_transcript_source_name\n",
       "28   percentage_gene_gc_content     \n",
       "29   gene_biotype                   \n",
       "30   transcript_biotype             \n",
       "⋮    ⋮                              \n",
       "3128 5_utr_end                      \n",
       "3129 3_utr_start                    \n",
       "3130 3_utr_end                      \n",
       "3131 ensembl_transcript_id          \n",
       "3132 ensembl_transcript_id_version  \n",
       "3133 ensembl_peptide_id             \n",
       "3134 ensembl_peptide_id_version     \n",
       "3135 transcript_biotype             \n",
       "3136 transcript_version             \n",
       "3137 peptide_version                \n",
       "3138 strand                         \n",
       "3139 transcript_start               \n",
       "3140 transcript_end                 \n",
       "3141 transcription_start_site       \n",
       "3142 transcript_length              \n",
       "3143 cds_length                     \n",
       "3144 cds_start                      \n",
       "3145 cds_end                        \n",
       "3146 ensembl_exon_id                \n",
       "3147 exon_chrom_start               \n",
       "3148 exon_chrom_end                 \n",
       "3149 strand                         \n",
       "3150 rank                           \n",
       "3151 phase                          \n",
       "3152 end_phase                      \n",
       "3153 cdna_coding_start              \n",
       "3154 cdna_coding_end                \n",
       "3155 genomic_coding_start           \n",
       "3156 genomic_coding_end             \n",
       "3157 is_constitutive                \n",
       "     description                                  page        \n",
       "1    Gene stable ID                               feature_page\n",
       "2    Gene stable ID version                       feature_page\n",
       "3    Transcript stable ID                         feature_page\n",
       "4    Transcript stable ID version                 feature_page\n",
       "5    Protein stable ID                            feature_page\n",
       "6    Protein stable ID version                    feature_page\n",
       "7    Exon stable ID                               feature_page\n",
       "8    Gene description                             feature_page\n",
       "9    Chromosome/scaffold name                     feature_page\n",
       "10   Gene start (bp)                              feature_page\n",
       "11   Gene end (bp)                                feature_page\n",
       "12   Strand                                       feature_page\n",
       "13   Karyotype band                               feature_page\n",
       "14   Transcript start (bp)                        feature_page\n",
       "15   Transcript end (bp)                          feature_page\n",
       "16   Transcription start site (TSS)               feature_page\n",
       "17   Transcript length (including UTRs and CDS)   feature_page\n",
       "18   Transcript support level (TSL)               feature_page\n",
       "19   GENCODE basic annotation                     feature_page\n",
       "20   APPRIS annotation                            feature_page\n",
       "21   Ensembl Canonical                            feature_page\n",
       "22   RefSeq match transcript (MANE Select)        feature_page\n",
       "23   RefSeq match transcript (MANE Plus Clinical) feature_page\n",
       "24   Gene name                                    feature_page\n",
       "25   Source of gene name                          feature_page\n",
       "26   Transcript name                              feature_page\n",
       "27   Source of transcript name                    feature_page\n",
       "28   Gene % GC content                            feature_page\n",
       "29   Gene type                                    feature_page\n",
       "30   Transcript type                              feature_page\n",
       "⋮    ⋮                                            ⋮           \n",
       "3128 5' UTR end                                   sequences   \n",
       "3129 3' UTR start                                 sequences   \n",
       "3130 3' UTR end                                   sequences   \n",
       "3131 Transcript stable ID                         sequences   \n",
       "3132 Transcript stable ID version                 sequences   \n",
       "3133 Protein stable ID                            sequences   \n",
       "3134 Protein stable ID version                    sequences   \n",
       "3135 Transcript type                              sequences   \n",
       "3136 Version (transcript)                         sequences   \n",
       "3137 Version (protein)                            sequences   \n",
       "3138 Strand                                       sequences   \n",
       "3139 Transcript start (bp)                        sequences   \n",
       "3140 Transcript end (bp)                          sequences   \n",
       "3141 Transcription start site (TSS)               sequences   \n",
       "3142 Transcript length (including UTRs and CDS)   sequences   \n",
       "3143 CDS Length                                   sequences   \n",
       "3144 CDS start                                    sequences   \n",
       "3145 CDS end                                      sequences   \n",
       "3146 Exon stable ID                               sequences   \n",
       "3147 Exon region start (bp)                       sequences   \n",
       "3148 Exon region end (bp)                         sequences   \n",
       "3149 Strand                                       sequences   \n",
       "3150 Exon rank in transcript                      sequences   \n",
       "3151 Start phase                                  sequences   \n",
       "3152 End phase                                    sequences   \n",
       "3153 cDNA coding start                            sequences   \n",
       "3154 cDNA coding end                              sequences   \n",
       "3155 Genomic coding start                         sequences   \n",
       "3156 Genomic coding end                           sequences   \n",
       "3157 Constitutive exon                            sequences   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## list the available datasets in this Mart\n",
    "listAttributes(mart = ensembl_mart)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gn <- biomaRt::getBM(filters=\"refseq_mrna\", \n",
    "                         attributes=c(\"refseq_mrna\",\"hgnc_symbol\",\"ensembl_gene_id\"), \n",
    "                         values=refseq_ids, mart=mart, useCache = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>refseq_mrna</th><th scope=col>hgnc_symbol</th><th scope=col>ensembl_gene_id</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>NM_001005337</td><td>PKP1</td><td>ENSG00000081277</td></tr>\n",
       "\t<tr><td>NM_012102   </td><td>RERE</td><td>ENSG00000142599</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " refseq\\_mrna & hgnc\\_symbol & ensembl\\_gene\\_id\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t NM\\_001005337 & PKP1 & ENSG00000081277\\\\\n",
       "\t NM\\_012102    & RERE & ENSG00000142599\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| refseq_mrna &lt;chr&gt; | hgnc_symbol &lt;chr&gt; | ensembl_gene_id &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| NM_001005337 | PKP1 | ENSG00000081277 |\n",
       "| NM_012102    | RERE | ENSG00000142599 |\n",
       "\n"
      ],
      "text/plain": [
       "  refseq_mrna  hgnc_symbol ensembl_gene_id\n",
       "1 NM_001005337 PKP1        ENSG00000081277\n",
       "2 NM_012102    RERE        ENSG00000142599"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref_gn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Session information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.4.0 (2024-04-24)\n",
       "Platform: x86_64-pc-linux-gnu\n",
       "Running under: CentOS Linux 7 (Core)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so;  LAPACK version 3.8.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: America/Denver\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       "[1] org.Hs.eg.db_3.19.1  AnnotationDbi_1.66.0 IRanges_2.38.0      \n",
       "[4] S4Vectors_0.42.0     Biobase_2.64.0       BiocGenerics_0.50.0 \n",
       "[7] biomaRt_2.60.0      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] rappdirs_0.3.3          utf8_1.2.4              generics_0.1.3         \n",
       " [4] xml2_1.3.6              RSQLite_2.3.7           stringi_1.8.4          \n",
       " [7] hms_1.1.3               digest_0.6.35           magrittr_2.0.3         \n",
       "[10] evaluate_0.23           pbdZMQ_0.3-11           fastmap_1.2.0          \n",
       "[13] blob_1.2.4              jsonlite_1.8.8          progress_1.2.3         \n",
       "[16] GenomeInfoDb_1.40.1     DBI_1.2.3               httr_1.4.7             \n",
       "[19] purrr_1.0.2             fansi_1.0.6             UCSC.utils_1.0.0       \n",
       "[22] Biostrings_2.72.1       httr2_1.0.1             cli_3.6.2              \n",
       "[25] rlang_1.1.4             crayon_1.5.2            dbplyr_2.5.0           \n",
       "[28] XVector_0.44.0          bit64_4.0.5             withr_3.0.0            \n",
       "[31] base64enc_0.1-3         repr_1.1.7              cachem_1.1.0           \n",
       "[34] tools_4.4.0             uuid_1.2-0              memoise_2.0.1          \n",
       "[37] dplyr_1.1.4             filelock_1.0.3          GenomeInfoDbData_1.2.12\n",
       "[40] curl_5.2.1              IRdisplay_1.1           vctrs_0.6.5            \n",
       "[43] R6_2.5.1                png_0.1-8               lifecycle_1.0.4        \n",
       "[46] BiocFileCache_2.12.0    zlibbioc_1.50.0         KEGGREST_1.44.0        \n",
       "[49] stringr_1.5.1           bit_4.0.5               pkgconfig_2.0.3        \n",
       "[52] pillar_1.9.0            glue_1.7.0              tidyselect_1.2.1       \n",
       "[55] tibble_3.2.1            IRkernel_1.3.2          htmltools_0.5.8.1      \n",
       "[58] compiler_4.4.0          prettyunits_1.2.0      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
